### PR TITLE
C make implementation

### PR DESCRIPTION
**Pull request by upadhyaydhruv**
CMake linking for all the code in the new AM architecture

Still under progress, marked as draft for now

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,6 +19,9 @@
 *.hex
 *.bin
 
+# Cache Files
+*.idx
```
Comment (upadhyaydhruv) [inline .gitignore:23]: This is to prevent cache from being committed every time xd

File: Common/CControl/Inc/CControlFunctions.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
+/*
```
Comment (upadhyaydhruv) [inline Common/CControl/Inc/CControlFunctions.h]: Migrated from Autopilot to common so sensorfusion on Safety can also access these functions

File: Common/CControl/Src/inv.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (upadhyaydhruv) [inline Common/CControl/Src/inv.c]: Migration to common

File: Common/Inc/AttitudePathInterface.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
+/**
```
Comment (upadhyaydhruv) [inline Common/Inc/AttitudePathInterface.hpp:1]: This contained a lot of information regading structs, data passed between AM and PM. Since PM is on Autopilot and AM on Safety, it was necessary to have this in common

File: Common/SensorFusion/Src/SensorFusion.cpp [hunk 5]
```diff
@@ -9,10 +9,12 @@
 #include <cmath>
 #include "MathConstants.hpp"
 #include "imu.hpp"
+#ifdef AUTOPILOT
```
Comment (upadhyaydhruv) [inline Common/SensorFusion/Src/SensorFusion.cpp:12]: @antimonyB added a bunch of header guards differentiating between sensorfusion on Autopilot and safety, since both have different sensors and requirements

File: Safety/AttitudeManager/Src/CommFromAMToPM.cpp [hunk 6]
```diff
@@ -1,65 +0,0 @@
-/**
```
Comment (upadhyaydhruv) [inline Safety/AttitudeManager/Src/CommFromAMToPM.cpp:1]: This needed to be deleted/rewritten since comms between AM and PM occur over interchip rather than mailQs now

File: Safety/AttitudeManager/Src/OutputMixing.cpp [hunk 7]
```diff
@@ -82,11 +82,11 @@ static OutputMixingErrorCodes checkInputValidity(PID_Output_t *PidOutput)
 {
 	OutputMixingErrorCodes errorCode;
 
-	if ( (PidOutput->frontLeft < -100.0f) || (PidOutput->pitchPercent < -100.0f) || (PidOutput->rudderPercent < -100.0f) || (PidOutput->throttlePercent < 0.0f) )
+	if ( (PidOutput->frontLeftMotorPercent < -100.0f) || (PidOutput->frontRightMotorPercent < -100.0f) || (PidOutput->backLeftMotorPercent < -100.0f) || (PidOutput->backRightMotorPercent < 0.0f) )
```
Comment (upadhyaydhruv) [inline Safety/AttitudeManager/Src/OutputMixing.cpp]: There was ambiguity between the tetrms used to refer to percentage output to PWM motors in the drone, has been fixed

File: Safety/CMakeLists.txt [hunk 8]
```diff
@@ -10,21 +10,30 @@ add_definitions(
   -DSTM32F030xC
 )
 
+get_filename_component(PARENT_DIR ../ ABSOLUTE)
+
 include_directories(
+    Inc
+    ${PARENT_DIR}/Common/Inc
```
Comment (upadhyaydhruv) [inline Safety/CMakeLists.txt:17]: Added so Safety can access common files

File: Safety/Inc/imu.hpp [hunk 9]
```diff
@@ -1,114 +1,49 @@
-/**
```
Comment (upadhyaydhruv) [inline Safety/Inc/imu.hpp:1]: Replaced BMX160 IMU driver with Neha's MPU6050 driver

**Comment by upadhyaydhruv** [comment]
In the interest of timelines I'm merging this for now

**Review by antholuo — APPROVED** [review]

File: Common/Inc/CControlFunctions.h [hunk 10]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Functions.h
+ *
+ *  Created on: 5 oct. 2019
+ *      Author: Daniel Mårtensson
+ */
+
```
Comment (Aidan-B) [inline Common/Inc/CControlFunctions.h:7]: Is there supposed to be two `CControlFunctions.h` files?

File: Common/CControl/Inc/CControlFunctions.h (hunk 2 above)
Comment (Aidan-B) [reply Common/CControl/Inc/CControlFunctions.h]: I don't see it removed from autopilot in this PR...?

File: Safety/AttitudeManager/Inc/AttitudeDatatypes.hpp [hunk 11]
```diff
@@ -37,5 +37,5 @@ Commented out for now to allow for just quadcopter channel mapping
 
 // constexpr static uint8_t NUM_PWM_CHANNELS {8};
 
-// #endif
+#endif
```
Comment (Aidan-B) [inline Safety/AttitudeManager/Inc/AttitudeDatatypes.hpp]: `#endif //ATTITUDE_DATATYPES_HPP`

File: Common/Inc/AttitudePathInterface.hpp [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+#ifdef TARGET_BUILD
+
+extern "C"
+{
+#include "cmsis_os.h"
+}
+
+const char PATH_ATTITUDE_MAIL_Q_SIZE = 1;
+
+#endif
```
Comment (Aidan-B) [inline Common/Inc/AttitudePathInterface.hpp:40]: `#endif /*  TARGET_BUILD */`

File: Common/Inc/AttitudePathInterface.hpp [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+#endif
```
Comment (Aidan-B) [inline Common/Inc/AttitudePathInterface.hpp:42]: `#endif /* ATTITUDE_PATH_INTERFACE_HPP */`

File: Common/Inc/CommonDataTypes.hpp [hunk 14]
```diff
@@ -0,0 +1,9 @@
+#ifndef COMMON_DATA_TYPES_HPP
+#define COMMON_DATA_TYPES_HPP
+
+enum SensorErrorCodes {SENSOR_SUCCESS = 0, SENSOR_FAIL = 1};
+
+
+#endif
```
Comment (Aidan-B) [inline Common/Inc/CommonDataTypes.hpp:7]: `#endif /* COMMON_DATA_TYPES_HPP */`

File: Common/SensorFusion/Inc/SensorFusion.hpp [hunk 15]
```diff
@@ -131,5 +133,6 @@ Gps_Data_t SF_GetRawGPS();
  * @return Altimeter struct.
  */ 
 Altimeter_Data_t SF_GetRawAltimeter();
+#endif
```
Comment (Aidan-B) [inline Common/SensorFusion/Inc/SensorFusion.hpp:136]: I feel like i'm getting repetitive with these `#endif`s... I'm just going to leave it be now even though there are more

File: Safety/AttitudeManager/Inc/attitudeStateClasses.hpp [hunk 16]
```diff
@@ -4,12 +4,12 @@
 #include "attitudeManager.hpp"
 #include "AttitudeDatatypes.hpp"
 #include "attitudeConfig.hpp"
+#include "../../Common/Inc/AttitudePathInterface.hpp"
+#include "../../Common/Inc/CommWithPathManager.hpp"
```
Comment (Aidan-B) [inline Safety/AttitudeManager/Inc/attitudeStateClasses.hpp]: Why did you need to step back two directories here... shouldn't the `Common/Inc/*` folder be found by Safety's build system?

File: Safety/AttitudeManager/Src/OutputMixing.cpp (hunk 7 above)
Comment (Aidan-B) [inline Safety/AttitudeManager/Src/OutputMixing.cpp]: Can we break these onto a few lines so that it's not so hard to read?

File: Safety/AttitudeManager/Src/OutputMixing.cpp [hunk 17]
```diff
@@ -82,11 +82,11 @@ static OutputMixingErrorCodes checkInputValidity(PID_Output_t *PidOutput)
...
 	{
 		errorCode = OUTPUT_MIXING_VALUE_TOO_LOW;
 	}
-	else if ( (PidOutput->rollPercent > 100.0f) || (PidOutput->pitchPercent > 100.0f) || (PidOutput->rudderPercent > 100.0f) || (PidOutput->throttlePercent > 100.0f) )
+	else if ( (PidOutput->frontLeftMotorPercent > 100.0f) || (PidOutput->frontRightMotorPercent > 100.0f) || (PidOutput->backLeftMotorPercent > 100.0f) || (PidOutput->backRightMotorPercent > 100.0f) )
```
Comment (Aidan-B) [inline Safety/AttitudeManager/Src/OutputMixing.cpp]: Can we break these onto a few lines so that it's not so hard to read?

File: Safety/Controls/Src/Controls.cpp [hunk 18]
```diff
@@ -9,7 +9,7 @@
 #include "../../Inc/PPM.hpp"
 #include "../../Inc/PWM.hpp"
 #include "base_pid.hpp"
-#include "../../SensorFusion/Inc/sensorFusion.hpp"
+#include "../../../Common/SensorFusion/Inc/SensorFusion.hpp"
```
Comment (Aidan-B) [inline Safety/Controls/Src/Controls.cpp]: Seems like the build system should also be able to find this without stepping back three folders...?

**Review by Aidan-B — COMMENTED** [review]
Few points:
- There are a number of `#endif`s that should probably be commented so it can be seen at a glance what it matches to.

- There also seem to be a lot of files that have been "moved" from autopilot to common, but I don't see the move or remove label in the review (So double check that everything has been moved in this PR as expected)

- There are a number of instances of `../../../Common/Inc/`... which seems like it should be handled by the build system to me.

- Finally, there are more than a few places where this PR seems to overstep from just Cmake linking. I don't think this blocks it, but I haven't really reviewed the logic of all the files. For example the Quaternion stuff. Maybe it's just because things are being moved around and I didn't clue in.

File: Safety/AttitudeManager/Inc/attitudeStateClasses.hpp (hunk 16 above)
Comment (upadhyaydhruv) [reply Safety/AttitudeManager/Inc/attitudeStateClasses.hpp]: You're right, that's fixed

File: Safety/AttitudeManager/Src/OutputMixing.cpp (hunk 7 above)
Comment (upadhyaydhruv) [reply Safety/AttitudeManager/Src/OutputMixing.cpp]: Will do

File: Common/CControl/Inc/CControlFunctions.h (hunk 2 above)
Comment (upadhyaydhruv) [reply Common/CControl/Inc/CControlFunctions.h]: You're right, should be removed now
